### PR TITLE
Added move constructor and move assignment operator for sf::Shader

### PR DESCRIPTION
**Pull request by Eren121**
* [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
Yes, multiple times in multiple topics
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [x] Have you provided some example/test code for your changes?
See related issue

**Related to >=C++11 change**
It's a syntactic change that should not break any cross-platform compatibility. The compatibility is related to the compiler and not the platform, and today most if not all modern compilers support C++11. But it can break C++03 code (any project use it?... if yes, they probably also use older versions of SFML that are also compatible with C++03).

Some tests did not pass because they aren't actually... C++11 compatible.

It's can also maybe be a change for another branch forecasting C++11.

----

## Description

Make sf::Shader moveable, and keep not copiable.

This PR is related to the issue https://github.com/SFML/SFML/issues/1675
Code from the issue:
```cpp
#include <SFML/Graphics.hpp>

sf::Shader loadShader() {
    sf::Shader shader;
    /* Probably more code here */
    return shader;
}
int main() {
    sf::Shader myShader = loadShader();
}
```

### CODE REVIEW

**Comment by unl1ght** [comment]
I noticed that the move assignment operator performs a swap of variables between the two instances. Would it be more appropriate to destroy the lhs object resources, take ownership of the rhs object resources, then reinitialise the rhs resources to default/null states?

**Comment by unl1ght** [comment]
I have some concerns with the use of swap in the move assignment operator that I have voiced on the forum. Please see https://en.sfml-dev.org/forums/index.php?topic=27116.msg173691#msg173691

File: include/SFML/Graphics/Shader.hpp [hunk 1]
```diff
@@ -91,12 +91,24 @@ class SFML_GRAPHICS_API Shader : GlResource, NonCopyable
     ////////////////////////////////////////////////////////////
     Shader();
 
+    ////////////////////////////////////////////////////////////
+    /// \brief Move constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    Shader(Shader&& other);
```
Comment (reconndev) [inline include/SFML/Graphics/Shader.hpp:98]: Move constructors should be marked `noexcept` whenever possible.

Also, looking at the implementation this seems like it should be:
```cpp
Shader(Shader&& other) noexcept = default;
```

File: include/SFML/Graphics/Shader.hpp [hunk 2]
```diff
@@ -91,12 +91,24 @@ class SFML_GRAPHICS_API Shader : GlResource, NonCopyable
...
     /// \brief Destructor
     ///
     ////////////////////////////////////////////////////////////
     ~Shader();
 
+    ////////////////////////////////////////////////////////////
+    /// \brief Move assignement
+    ///
+    ////////////////////////////////////////////////////////////
+    Shader& operator=(Shader&& other);
```
Comment (reconndev) [inline include/SFML/Graphics/Shader.hpp:110]:
```cpp
Shader& operator=(Shader&& other) noexcept = default;
```

File: src/SFML/Graphics/Shader.cpp [hunk 3]
```diff
@@ -228,6 +229,14 @@ m_uniforms      ()
 }
 
 
+////////////////////////////////////////////////////////////
+Shader::Shader(Shader&& other) :
+Shader()
+{
+    *this = std::move(other);
```
Comment (reconndev) [inline src/SFML/Graphics/Shader.cpp:236]: You're using operator assignment inside constructor logic which loses the benefits of in-place construction.

Solution: Definitions are not needed when the declaration states `= default`

File: src/SFML/Graphics/Shader.cpp [hunk 4]
```diff
@@ -239,6 +248,22 @@ Shader::~Shader()
 }
 
 
+////////////////////////////////////////////////////////////
+Shader& Shader::operator=(Shader&& other)
```
Comment (reconndev) [inline src/SFML/Graphics/Shader.cpp:252]: Same as before: should probably be removed

**Comment by Bromeon** [comment]
Valid points. There are different ways to implement move semantics in C++, we should agree on one.

I brought this up in the forum a while ago, see [here](https://en.sfml-dev.org/forums/index.php?topic=27116.msg173694#msg173694) and subsequent posts.

**Comment by reconndev** [comment]
> There are different ways to implement move semantics in C++, we should agree on one.

I think it is important to keep a strong exception guarantee for move semantics. The transfer should always be intuitive, and I believe that `noexcept` grants this intuitiveness. I also posted my view on the forum thread you mentioned.

**Comment by JonnyPtn** [comment]
superseded by #2277 ?

**Comment by eXpl0it3r** [comment]
Yep ^

**Pull request closed without merging**